utils/providers.py

```diff
@@ -176,13 +177,24 @@ def setup_a_provider(
             # If our filtering yielded any providers, use them, otherwise do not bother with that
             providers = filtered_providers
 
+    # Keep track of providers that do not behave correctly
+    bad_providers = set([])
     # If there is already a suitable provider, don't try to setup a new one.
     already_existing = filter(is_provider_setup, providers)
-    if already_existing:
-        chosen = random.choice(already_existing)
+    random.shuffle(already_existing)
```

Comment:
If the providers behave correctly, then it is going to validate only one of them. In case that one is no good, then it checks another.